src/java/htsjdk/samtools/SAMRecord.java

```diff
@@ -342,26 +343,19 @@ private boolean hasMateReferenceName() {
      * the reference index to NO_ALIGNMENT_REFERENCE_INDEX.
      *
      * @param referenceName - must not be null
+     * @throws IllegalArgumentException if {@code referenceName} is null
      */
     public void setReferenceName(final String referenceName) {
         if (null == referenceName) {
             throw new IllegalArgumentException(
                     "Reference name must not be null. Use SAMRecord.NO_ALIGNMENT_REFERENCE_NAME to reset the reference name.");
         }
-        else if (NO_ALIGNMENT_REFERENCE_NAME.equals(referenceName)) {
-            mReferenceName = NO_ALIGNMENT_REFERENCE_NAME;
-            mReferenceIndex = NO_ALIGNMENT_REFERENCE_INDEX;
```

Comment:
Previously if `NO_ALIGNMENT_REFERENCE_NAME` was passed in here, `mReferenceIndex` would get set to `NO_ALIGNMENT_REFERENCE_INDEX` even if the header was null. Now if the header is null, `mReferenceIndex` will get set to null in this case. Is there a reason for this behavioral change? Ideally we shouldn't change any existing behavior when refactoring to use the new `resolve*` methods.